pkg/util/option_values.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2024 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package util
+
+import (
+	"fmt"
+	"reflect"
+
+	vimtypes "github.com/vmware/govmomi/vim25/types"
+)
+
+// OptionValues simplifies manipulation of properties that are arrays of
+// vimtypes.BaseOptionValue, such as ExtraConfig.
+type OptionValues map[string]string
+
+// OptionValuesFromList returns a new OptionValues object from the provided
+// list.
+func OptionValuesFromList(in []vimtypes.BaseOptionValue) OptionValues {
+	if len(in) == 0 {
+		return nil
+	}
+	out := make(OptionValues, len(in))
+	for i := range in {
+		if optVal := in[i].GetOptionValue(); optVal != nil {
+			out[optVal.Key] = getOptionValue(*optVal)
+		}
+	}
+	return out
+}
+
+// ToList returns a list of vimtypes.BaseOptionValue objects.
+func (ov OptionValues) ToList() []vimtypes.BaseOptionValue {
+	if len(ov) == 0 {
+		return nil
+	}
+	i := 0
+	out := make([]vimtypes.BaseOptionValue, len(ov))
+	for k, v := range ov {
+		out[i] = &vimtypes.OptionValue{Key: k, Value: v}
+		i++
+	}
+	return out
+}
+
+// Diff returns the difference between the existing data and provided input.
+// Please note the output does not maintain the order of the input.
+func (ov OptionValues) Diff(in OptionValues) OptionValues {
+	out := OptionValues{}
+	for k, newVal := range in {
+		if oldVal, ok := ov[k]; !ok || newVal != oldVal {
+			out[k] = newVal
+		}
+	}
+	if len(out) == 0 {
+		return nil
```

Comment:
Depending on how these calls are all chain, this does worry me a little that we can unexpectedly NPE, like if somebody would then call a func that ends up in `merge()` or `mergeList()`

Reply:
How so? We allow `in` to be `nil` without error, and the functions that could NPE do not return anything, such as `Merge` and `MergeList`. These specifically do not return and cannot be used to chain calls.

Reply:
Let me rephrase a little bit - with the type alias you still need need to know the underlying type is nullable. I'm worried that something like below is too easy to write or end up with:

```
in := []vimtypes.BaseOptionValue{&vimtypes.OptionValue{Key: "foo", Value: "one"}}
more := []vimtypes.BaseOptionValue{&vimtypes.OptionValue{Key: "foo1", Value: "two"}}
x := pkgutil.OptionValuesFromList(in).DiffList(in)
....
x.MergeList(more)
```

Thinking back to a bug that was introduced with the switch to either the `slices` or `maps` package that an empty in param return nil.